Race condition in instance.GetWmiInstanceManager

The `instanceManagerMap` is guarded using a mutex to ensure concurrent usages. However, the mutex is not locked during accessing the map, causing race condition.

I don't have a windows machine, but this code mimics the GetWmiInstanceManager code:

```
package main

import (
	"sync"
	"testing"
)

var (
	m     map[int]*int
	mutex sync.Mutex
)

func init() {
	m = map[int]*int{}
}

func f(id int) *int {
	if val, ok := m[id]; ok {
		return val
	}

	mutex.Lock()
	defer mutex.Unlock()
	m[id] = nil
	return m[id]
}

func TestF(t *testing.T) {
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		f(1)
	}()
	go func() {
		defer wg.Done()
		f(1)
	}()
	wg.Wait()
}
```

Run test with `-race`:

```
go1.24.2 test -race main_test.go 
==================
WARNING: DATA RACE
Read at 0x00c000126570 by goroutine 9:
  runtime.mapaccess1()
      /home/cuonglm/sdk/go1.24.2/src/internal/runtime/maps/runtime_swiss.go:43 +0x0
  command-line-arguments.f()
      /home/cuonglm/main_test.go:18 +0x56
  command-line-arguments.TestF.func2()
      /home/cuonglm/main_test.go:37 +0x76

Previous write at 0x00c000126570 by goroutine 8:
  runtime.mapassign()
      /home/cuonglm/sdk/go1.24.2/src/internal/runtime/maps/runtime_swiss.go:191 +0x0
  command-line-arguments.f()
      /home/cuonglm/main_test.go:24 +0xf3
  command-line-arguments.TestF.func1()
      /home/cuonglm/main_test.go:33 +0x76

Goroutine 9 (running) created at:
  command-line-arguments.TestF()
      /home/cuonglm/main_test.go:35 +0x12a
  testing.tRunner()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1851 +0x44

Goroutine 8 (running) created at:
  command-line-arguments.TestF()
      /home/cuonglm/main_test.go:31 +0xc4
  testing.tRunner()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1851 +0x44
==================
==================
WARNING: DATA RACE
Read at 0x00c00008e010 by goroutine 9:
  command-line-arguments.f()
      /home/cuonglm/main_test.go:18 +0x64
  command-line-arguments.TestF.func2()
      /home/cuonglm/main_test.go:37 +0x76

Previous write at 0x00c00008e010 by goroutine 8:
  command-line-arguments.f()
      /home/cuonglm/main_test.go:24 +0x104
  command-line-arguments.TestF.func1()
      /home/cuonglm/main_test.go:33 +0x76

Goroutine 9 (running) created at:
  command-line-arguments.TestF()
      /home/cuonglm/main_test.go:35 +0x12a
  testing.tRunner()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1851 +0x44

Goroutine 8 (running) created at:
  command-line-arguments.TestF()
      /home/cuonglm/main_test.go:31 +0xc4
  testing.tRunner()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /home/cuonglm/sdk/go1.24.2/src/testing/testing.go:1851 +0x44
==================
--- FAIL: TestF (0.00s)
    testing.go:1490: race detected during execution of test
FAIL
FAIL	command-line-arguments	0.007s
FAIL
```

--- a/pkg/base/instance/instancemanager.go
+++ b/pkg/base/instance/instancemanager.go
@@ -54,20 +54,20 @@ func GetWmiInstanceManagerFromCred(hostname, namespaceName string, cred *credent
 	return GetWmiInstanceManager(hostname, namespaceName, cred.UserName, cred.Password, cred.Domain)
 }
 func GetWmiInstanceManager(hostname, namespaceName, userName, password, domainName string) (*WmiInstanceManager, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
 	mapId := strings.Join([]string{hostname, namespaceName, domainName}, "_")
 	if val, ok := instanceManagerMap[mapId]; ok {
 		return val, nil
 	}
 
-	mutex.Lock()
-	defer mutex.Unlock()
 	var err error
 	instanceManagerMap[mapId], err = newWmiInstanceManager(hostname, namespaceName, userName, password, domainName)
 	if err != nil {
 		return nil, err
 	}
 	return instanceManagerMap[mapId], nil
-
 }
 
 func (im *WmiInstanceManager) CreateInstance(className string) (*wmi.WmiInstance, error) {

--- a/pkg/base/instance/instancemanager_test.go
+++ b/pkg/base/instance/instancemanager_test.go
@@ -1,0 +1,23 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package instance
+
+import (
+	"strconv"
+	"sync"
+	"testing"
+)
+
+func TestGetWmiInstanceManagerConcurrently(t *testing.T) {
+	n := 2
+	var wg sync.WaitGroup
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func(name string) {
+			defer wg.Done()
+			GetWmiInstanceManager(name, "", "", "", "")
+		}(strconv.Itoa(n))
+	}
+	wg.Wait()
+}

